functions/Get-CSVMetadata.ps1

```diff
@@ -0,0 +1,283 @@
+            $HeaderRowParser.Close()
+            $HeaderRowParser.Dispose()
+            $sr.Close()
+            $sr.Dispose()    
+        }
+
+        if ($SampleRow)
+        {
+            $sr = New-Object System.IO.StringReader($SampleRow)
+            $SampleRowParser = New-Object Microsoft.VisualBasic.FileIO.TextFieldParser($sr)
+            $SampleRowParser.HasFieldsEnclosedInQuotes = $HasFieldsEnclosedInQuotes
+            $SampleRowParser.Delimiters = $delimiter
+            $RowColumns = $SampleRowParser.ReadFields()
+            $SampleRowParser.Close()
+            $SampleRowParser.Dispose()
+            $sr.Close()
+            $sr.Dispose()    
+        }
+
+        $results = @()
+
+        if ($Headers)
+        {
+            if($HeaderColumns.Length -eq $RowColumns.Length)
+            {
+                Write-Verbose ("[$FileName] Header and sample row column count matching H:{0} R:{1}" -f $HeaderColumns.Length , $RowColumns.Length)
+            }
+            elseif($HeaderColumns.Length -gt $RowColumns.Length)
+            {
+                Write-Warning ("[$FileName] More header than row column. H:{0} R:{1}" -f $HeaderColumns.Length , $RowColumns.Length)
+                Write-Warning "[$FileName] Stripping out headers."
+                $HeaderColumns = $null
+                Get-UserPrompt
+            }
+            else
+            {   
+                Write-Warning ("[$FileName] More header than row column. H:{0} R:{1} unable to match" -f $HeaderColumns.Length , $RowColumns.Length)
+                Write-Warning "[$FileName] Stripping out headers."
+                Get-UserPrompt
+                $HeaderColumns = $null                 
+            }
+        }
+        else
+        {
+            Write-Verbose "[$FileName] No header specified, skipping header and sample row column comparison." 
+        }
+        
+        Write-Verbose "[$FileName] Generating metadata (Header name, length, data type matrix)."
+        
+        foreach ($column in $RowColumns) 
+        {
+            try 
+            {                
+                $HeaderName = $HeaderColumns[$RowColumns.indexof($column)]
+            } 
+            catch 
+            {
+                $HeaderName = ("Column" + ($RowColumns.indexof($column) + 1))
+            }
+
+            #  [char]      A Unicode 16-bit character
+            #  [byte]      An 8-bit unsigned character
+            #  [int]       32-bit signed integer
+            #  [long]      64-bit signed integer
+            #  [bool]      Boolean True/False value
+            #  [decimal]   A 128-bit decimal value
+            #  [single]    Single-precision 32-bit floating point number
+            #  [double]    Double-precision 64-bit floating point number
+            #  [DateTime]  Date and Time
+            #  [string]    Fixed-length string of Unicode characters
+    
+            $hash = New-Object System.Object    

```

Comment:
Please use [pscustomobject] instead. It's [beautiful and faster](http://www.jonathanmedd.net/2011/09/powershell-v3-creating-objects-with-pscustomobject-its-fast.html)

    [PSCustomObject]@{
        HeaderName = $HeaderName
        Column = $Column
        SampleLength = $SampleLength 
    }

Piping should be avoided in this case because it's slower